File: src/ParallelFuzzyMatch/ParallelPatterns.FuzzyMatch/Module2/Pipeline.cs
using System;
using System.Collections.Concurrent;
using System.Diagnostics.Contracts;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using ParallelPatterns.Fsharp;

namespace ParallelPatterns
{
    public class Pipeline<TInput, TOutput>  
    {
        private readonly Func<TInput, Task<TOutput>> _processoTask;
        private readonly Func<TInput, TOutput> _processor;
        private readonly BlockingCollection<TInput>[] _input;
        private readonly CancellationToken _token;
        private const int Count = 3;

        private Pipeline(
            Func<TInput, Task<TOutput>> processor,
            BlockingCollection<TInput>[] input = null,
            CancellationToken token = new CancellationToken())
        {
            _input = input ?? Enumerable.Range(0, Count - 1).Select(_ => new BlockingCollection<TInput>(10)).ToArray();
            
            Output = new BlockingCollection<TOutput>[_input.Length];
            for (var i = 0; i < Output.Length; i++)
                Output[i] = null == _input[i] ? null : new BlockingCollection<TOutput>(Count);
            
            _processoTask = processor;
            _token = token;
            Task.Factory.StartNew(Run, _token, TaskCreationOptions.LongRunning, TaskScheduler.Default);
        }

        private Pipeline(
            Func<TInput, TOutput> processor,
            BlockingCollection<TInput>[] input = null,
            CancellationToken token = new CancellationToken())
        {
            _input = input ?? Enumerable.Range(0, Count - 1).Select(_ => new BlockingCollection<TInput>(10)).ToArray();
            
            Output = new BlockingCollection<TOutput>[_input.Length];
            for (var i = 0; i < Output.Length; i++)
                Output[i] = null == _input[i] ? null : new BlockingCollection<TOutput>(Count);
            
            _processor = processor;
            _token = token;
            Task.Factory.StartNew(Run, _token, TaskCreationOptions.LongRunning, TaskScheduler.Default);
        }

        public static Pipeline<TInput, TOutput> Create(
            Func<TInput, TOutput> processor,
            CancellationToken token = new CancellationToken())
            => new Pipeline<TInput, TOutput>(processor, token: token);

        public static Pipeline<TInput, TOutput> Create(
            Func<TInput, Task<TOutput>> processor,
            CancellationToken token = new CancellationToken()) 
            => new Pipeline<TInput, TOutput>(processor, token: token);
        
        private BlockingCollection<TOutput>[] Output { get; }

        public Pipeline<TOutput, TMid> Then<TMid>(
            Func<TOutput, Task<TMid>> project,
            CancellationToken token = new CancellationToken())
            => new Pipeline<TOutput, TMid>(project, Output, token);

        public Pipeline<TOutput, TMid> Then<TMid>(
            Func<TOutput, TMid> project,
            CancellationToken token = new CancellationToken())
            => new Pipeline<TOutput, TMid>(project, Output, token);
        
        // TODO (3.a)
        public void Enqueue(TInput item)
        {
            // Missing code
            // add item to internal buffer
        }

        private async Task Run()
        {
            var sw = new SpinWait();
            while (!_input.All(bc => bc.IsCompleted) && !_token.IsCancellationRequested)
            {
                // TODO (3.b)
                // Missing code
                // step to implement 
                // 1 - take an item from the _input collection 

                // 2 - process the item with the internal function
                //          either _processor or _processoTask according to the active one
                // 3 - push the result to the Output collec
                // Bouns :  avoid contention in case of empty queue. 
                //          Check "SpinWait" (see aboue instance "sw")
            }  

            if (Output != null)
            {
                foreach (var bc in Output) bc.CompleteAdding();
            }
        }
    }
}

File: src/ParallelPatterns/src/Module2/Pipeline.FSharp/Pipeline.fs
namespace Pipeline.FSharp

open System
open System.Linq
open System.Collections.Concurrent
open System.Threading
open System.Threading.Tasks


module PipelineHelpers =

    // This is a version of the pipeline that uses the .NET Func delegate
    // in place of the F# functions.
    // This implementation is more C# friendly, because it does not require you to
    // convert the F# functions to Func<> using the helper extension method .ToFunc()
    let composeFunc (f1:Func<_, _>) (f2:Func<_, _>) = Func<_,_>(fun x -> f2.Invoke(f1.Invoke(x)))

    let composeTasks (input : Task<'T>) ( binder :Func<'T, Task<'U>>) =
            let tcs = new TaskCompletionSource<'U>()
            input.ContinueWith(fun (task:Task<'T>) ->
               if (task.IsFaulted) then
                    tcs.SetException(task.Exception.InnerExceptions)
               elif (task.IsCanceled) then tcs.SetCanceled()
               else
                    try
                       (binder.Invoke(task.Result)).ContinueWith(fun(nextTask:Task<'U>) -> tcs.SetResult(nextTask.Result)) |> ignore
                    with
                    | ex -> tcs.SetException(ex)) |> ignore
            tcs.Task

open PipelineHelpers
// IPipeline interface
[<Interface>]
type IPipeline<'a,'b> =
    abstract member Then : (Func<'b, 'c>) -> IPipeline<'a, 'c>
    abstract member Then : (Func<'b, Task<'c>>) -> IPipeline<'a, 'c>
    abstract member Enqueue : 'a * (Func<('a * 'b), unit>) -> unit
    abstract member Execute : int * CancellationToken -> IDisposable
    abstract member Stop : unit -> unit

[<Struct>]
type internal Continuation<'a, 'b>(input:'a, callback:Func<('a * 'b), unit>) =
    member this.Input with get() = input
    member this.Callback with get() = callback

// The Parallel Functional Pipeline pattern
type Pipeline<'a, 'b> private (func:Func<'a, 'b> option, funcTask:Func<'a, Task<'b>> option) =
    let continuations = Array.init 3 (fun _ -> new BlockingCollection<Continuation<'a,'b>>(100))

    let func = defaultArg func (Func<_,_>(fun x -> Unchecked.defaultof<_>))
    let funcTask = defaultArg funcTask (Func<_,_>(fun x -> Unchecked.defaultof<_>))

    let then' (nextFunction:Func<'b,'c>) =
        Pipeline( (Some(composeFunc func nextFunction)), None) :> IPipeline<_,_>

    // TODO (3.a)
    let thenTask (nextFunction:Func<'b, Task<'c>>) =
        let task arg = funcTask.Invoke(arg)
        Pipeline(None, Some(Func<'a, Task<'c>>(fun arg -> composeTasks (task arg) nextFunction))) :> IPipeline<_,_>

    // TODO (3.b)
    let enqueue (input:'a) (callback:Func<('a * 'b), unit>) =
        // missing code
        ()

    let stop() = for continuation in continuations do continuation.CompleteAdding()

    let execute blockingCollectionPoolSize (cancellationToken:CancellationToken) =

        cancellationToken.Register(Action(stop)) |> ignore

        for i = 0 to blockingCollectionPoolSize - 1 do
            Task.Factory.StartNew(fun ( )->
                while (not <| continuations.All(fun bc -> bc.IsCompleted)) && (not <| cancellationToken.IsCancellationRequested) do
                        
                    // TODO (3.c)
                    // step to implement 
                        // 1 - take an item from the continuations collection 
                        // 2 - process the "continuation" function
                        //     Keep in mind that the continutaion function has both the
                        //     value and the callback 
                      
                    // let continuation = ref Unchecked.defaultof<Continuation<_,_>>
                    ()
            , cancellationToken, TaskCreationOptions.LongRunning, TaskScheduler.Default) |> ignore

    static member Create(func:Func<'a, 'b>) = Pipeline(Some(func), None) :> IPipeline<_,_>
    static member Create(func:Func<'a, Task<'b>>) = Pipeline(None, Some(func)) :> IPipeline<_,_>

    interface IPipeline<'a, 'b> with
        member this.Then(nextFunction) = then' nextFunction
        member this.Then(nextFunction:Func<'b, Task<'c>>) = thenTask nextFunction
        member this.Enqueue(input, callback) = enqueue input callback |> ignore
        member this.Stop() = stop()
        member this.Execute (blockingCollectionPoolSize, cancellationToken) =
            execute blockingCollectionPoolSize cancellationToken
            { new IDisposable with member self.Dispose() = stop() }
